### PR TITLE
Np 581 read data from crossref

### PR DESCRIPTION
**Pull request by axthosarouris**

### CODE REVIEW

File: config/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -256,11 +256,12 @@
                     LITERAL_SWITCH, LITERAL_SYNCHRONIZED, LITERAL_WHILE, METHOD_CALL,
                     METHOD_DEF, QUESTION, RESOURCE_SPECIFICATION, SUPER_CTOR_CALL, LAMBDA"/>
     </module>
+    <!-- Put Operators at the end of the line -->
     <module name="OperatorWrap">
-      <property name="option" value="NL"/>
       <property name="tokens"
-                value="BAND, BOR, BSR, BXOR, DIV, EQUAL, GE, GT, LAND, LE, LITERAL_INSTANCEOF, LOR,
-                    LT, MINUS, MOD, NOT_EQUAL, PLUS, QUESTION, SL, SR, STAR, METHOD_REF "/>
+        value="ASSIGN,DIV_ASSIGN,PLUS_ASSIGN,MINUS_ASSIGN,STAR_ASSIGN,MOD_ASSIGN,
```
Comment (axthosarouris) [inline config/checkstyle/checkstyle.xml]: Put logical operators at the end of line instead of at the beginning of the next line

File: src/main/java/no/unit/nva/doi/DataciteContentType.java [hunk 2]
```diff
@@ -23,22 +23,24 @@
...
-                .orElseThrow(() ->
-                    new IllegalArgumentException(
-                            String.format("Datacite Content Type not found for '%s', expected one of '%s'.",
-                                    contentType,
-                                    String.join(",", Arrays
-                                            .stream(DataciteContentType.values())
-                                            .map(DataciteContentType::getContentType)
-                                            .collect(Collectors.joining(",")))))
-                );
+        return Arrays
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/DataciteContentType.java]: Automatic linting

File: src/main/java/no/unit/nva/doi/DoiLookup.java [hunk 3]
```diff
@@ -1,17 +1,15 @@
 package no.unit.nva.doi;
 
-import java.net.URL;
-
 public class DoiLookup {
 
-    private URL doiUrl;
+    private String doi;
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/DoiLookup.java]: doi can be both a URL or a DOI string

File: src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            System.out.print(e.getMessage());
+            logger.log(e.getMessage());
+            return Optional.empty();
+        }
+
+    }
+
+    private HttpRequest createRequest(URI doiUri) {
+        return HttpRequest.newBuilder(doiUri)
+                          .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON)
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/CrossRefClient.java]: Request headers will be fixed on another PR

File: src/main/java/no/unit/nva/doi/FetchDoiMetadata.java [hunk 5]
```diff
@@ -26,40 +27,49 @@
 
     public static final String INVALID_DOI_URL = "The property 'doi_url' is not a valid DOI";
     public static final String MISSING_ACCEPT_HEADER = "Missing Accept header";
-    private static final Pattern DOI_URL_PATTERN = Pattern.compile("^https?://(dx\\.)?doi\\.org/"
-            + "(10(?:\\.[0-9]+)+)/(.+)$", Pattern.CASE_INSENSITIVE);
+    public static final Pattern DOI_URL_PATTERN =
+        Pattern.compile("^https?://(dx\\.)?doi\\.org/(10(?:\\.[0-9]+)+)/(.+)$",
+                        Pattern.CASE_INSENSITIVE);
+
+    public static final Pattern DOI_STRING_PATTERN =
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/FetchDoiMetadata.java]: Accept doi strings with or without the "doi:" prefix

File: src/main/java/no/unit/nva/doi/FetchDoiMetadata.java [hunk 6]
```diff
@@ -69,32 +79,47 @@ public GatewayResponse handleRequest(Map<String, Object> input, Context context)
...
+        this.logger = context.getLogger();
+        this.crossRefClient.setLogger(logger);
+    }
+
+    private String lookupDoiMetadata(String doiUrl, DataciteContentType dataciteContentType)
+        throws IOException, InterruptedException, ExecutionException, URISyntaxException {
         System.out.println("getDoiMetadata(doi:" + doiUrl + ")");
-        return dataciteClient.fetchMetadata(doiUrl.toString(), dataciteContentType);
+        Optional<String> crossRefResult = crossRefClient.fetchDataForDoi(doiUrl);
+        if (crossRefResult.isEmpty()) {
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/FetchDoiMetadata.java]: tried to chain them, but had problems with the exception handling. I have a suggestion for this that will allow us to have a more fluent syntax in the future

File: build.gradle [hunk 7]
```diff
@@ -9,21 +9,26 @@ plugins {
...
 dependencies {
     implementation group: 'org.apache.commons', name: 'commons-lang3', version: '3.9'
-    implementation group: 'javax.ws.rs', name: 'javax.ws.rs-api', version: '2.1.1'
+    implementation 'org.apache.httpcomponents:httpclient:4.5.12'
+    implementation group: 'org.glassfish.jersey.core', name: 'jersey-common', version: '2.30.1'
     implementation group: 'com.google.code.gson', name: 'gson', version: '2.8.6'
     implementation group: 'org.slf4j', name: 'slf4j-api', version: '1.7.29'
     implementation group: 'org.jsoup', name: 'jsoup', version: '1.12.1'
+
+    implementation 'com.github.BIBSYSDEV:aws-build-tools:v1.5.5'
```
Comment (janbj) [inline build.gradle]: Use of tag, excellent!

File: src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        } else {
+            return handleError(request, response);
+        }
+    }
+
+    private String handleError(HttpRequest request, HttpResponse<String> response) {
+        if (response.statusCode() == HttpStatus.SC_NOT_FOUND) {
+            throw new NotFoundException(COULD_NOT_FIND_ENTRY_WITH_DOI + request.uri().toString());
+        }
+        throw new BadRequestException("Something went wrong");
```
Comment (janbj) [inline src/main/java/no/unit/nva/doi/CrossRefClient.java]: Should we log the other statusCode or something from the respons body?

File: src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    private boolean responseIsSuccessful(HttpResponse<String> response) {
+        return response.statusCode() == HttpStatus.SC_OK;
```
Comment (janbj) [inline src/main/java/no/unit/nva/doi/CrossRefClient.java]: Are we satisfied with just 200 or should we check against the whole 2XX family?

File: src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    protected URI createTargetUrl(String doiIdentifier)
```
Comment (janbj) [inline src/main/java/no/unit/nva/doi/CrossRefClient.java]: What is target?

File: src/test/java/no/unit/nva/doi/FetchDoiMetadataTest.java [hunk 11]
```diff
@@ -33,43 +41,54 @@
...
+    public static final String VALID_DOI_WITHOUT_DOI_PREFIX = "10.123.4.5/124";
+
+    private static final Context mockLambdaContext = createMockContext();
+    private static final CrossRefClient crossRefClient = createCrossRefClient();
+
     public static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
 
     @Test
     public void successfulResponse() throws Exception {
         DataciteClient dataciteClient = Mockito.mock(DataciteClient.class);
```
Comment (janbj) [inline src/test/java/no/unit/nva/doi/FetchDoiMetadataTest.java]: Can be moved to class level and annotated with @Mock

File: src/test/java/no/unit/nva/utils/MockHttpClient.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+import java.net.http.HttpResponse.BodyHandler;
+import java.net.http.HttpResponse.PushPromiseHandler;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLParameters;
+
+public class MockHttpClient<R> extends HttpClient {
```
Comment (janbj) [inline src/test/java/no/unit/nva/utils/MockHttpClient.java]: This is a lot of code. Normal mocking wasn't sufficient?

File: template.yaml [hunk 13]
```diff
@@ -106,7 +106,7 @@ Resources:
     Properties:
       CodeUri: .
       Handler: no.unit.nva.doi.FetchDoiMetadata::handleRequest
-      Runtime: java8
+      Runtime: java11
```
Comment (janbj) [inline template.yaml]: Great!

File: src/main/java/no/unit/nva/doi/CrossRefClient.java (hunk 9 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/CrossRefClient.java]: It is a GET but I can extend it to check for all 2XX

File: src/test/java/no/unit/nva/utils/MockHttpClient.java (hunk 12 above)
Comment (axthosarouris) [reply src/test/java/no/unit/nva/utils/MockHttpClient.java]: I am not sure how to create a generic Mock.

File: src/main/java/no/unit/nva/doi/CrossRefClient.java (hunk 8 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/CrossRefClient.java]: Fixed

File: src/main/java/no/unit/nva/doi/CrossRefClient.java (hunk 9 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/CrossRefClient.java]: Fixed

File: src/main/java/no/unit/nva/doi/CrossRefClient.java (hunk 10 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/CrossRefClient.java]: Fixed

File: src/test/java/no/unit/nva/doi/FetchDoiMetadataTest.java (hunk 11 above)
Comment (axthosarouris) [reply src/test/java/no/unit/nva/doi/FetchDoiMetadataTest.java]: To my understanding the `@Mock` tag replaces the `mock` command. The methods have some `when` in the setups. I am not sure I can have the tag and keep the methods as they are now. If I keep the tags, then I would have to create an `@BeforeAll` method to initialize them, which becomes more complex than field initialization.
